File: leetcode-solutions/ruby/130. Surrounded Regions.rb
# https://leetcode.com/problems/surrounded-regions/
# we will flip all the 'O' which do not have any bounding 'O' neighbours
# in other work, we will keep those 'O' which has bounding 'O' neighbours

# @param {Character[][]} board
# @return {Void} Do not return anything, modify board in-place instead.
def solve(board)
    return if board.size.zero?
    
    # find all the 'O' which has bound neighbours and mark with 1

    # first row
    row = 0
    for col in 0...board.first.size
        next if board[row][col] != 'O'
        bfs(board, row, col)    
    end
    
    # last row
    row = board.size - 1
    for col in 0...board.first.size
        next if board[row][col] != 'O'
        bfs(board, row, col)    
    end
    
    # first col
    col = 0
    for row in 0...board.size
        next if board[row][col] != 'O'
        bfs(board, row, col)    
    end
    
    # last col
    col = board.first.size - 1
    for row in 0...board.size
        next if board[row][col] != 'O'
        bfs(board, row, col)    
    end
    
    # if the value is 1 that means that has bounding 'O' neighbours
    # so, that will be 'O' or convernt 'O' to 'x'
    for i in 0...board.size
        for j in 0...board.first.size
                board[i][j] = 'O'
            elsif board[i][j] == 'O'
                board[i][j] = 'X'
            end
        end
    end
    
    board        
end

def bfs(board, i, j)
    q = [[i, j]]
    max_row = board.size - 1
    max_col = board.first.size
    neighbours = [[-1, 0], [0,-1], [0, 1], [1, 0]]
    board[i][j] = 1
    while q.any?
        x, y = q.pop
        neighbours.each do |r, c|
            if x + r >= 0 && x + r <= max_row && y + c >=0 && y + c <= max_col && board[x+r][y+c] == 'O'
                board[x+r][y+c] = 1
                q << [x+r, y+c]
            end
                
        end
    end
end

File: leetcode-solutions/python/130. Surrounded Regions.py
# https://leetcode.com/problems/surrounded-regions/

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == 'O':
                    self.flip(board, i, j)
        
    def flip(self, board, i, j):
        m_row = 0
        m_rol= 0
        m_row = len(board) - 1
        m_col = len(board[0]) - 1
        ans = []
        visited = {}
        # mark visited
        visited[str(i) + str(j)] = True
        
        # keep as row, col         
        stack = [[i, j]]
        on_edge = False
        n_points = [
            [0, -1],
            [0, 1],
            [-1, 0],
            [1, 0]
        ]
        while stack:
            current = stack.pop()
            ans.append(current)
            # check element on the edge or not
            if current[0] == 0 or current[1] == 0 or current[0] == m_row or current[1] == m_col:
                on_edge = True
            
            # pick 4 neighbout
            for i, points in enumerate(n_points):
                n_row = current[0] + points[0]
                n_col = current[1] + points[1]
                key = str(n_row) + str(n_col)
                if n_row >= 0 and n_row <= m_row and n_col >= 0 and n_col <= m_col and key not in visited and board[n_row][n_col] == 'O':
                    # mark visited
                    visited[str(n_row) + str(n_col)] = True
                    stack.append([current[0] + points[0], current[1] + points[1]])
        if not on_edge:
            for i, points in enumerate(ans):
                board[points[0]][points[1]] = 'X'
            
        